### PR TITLE
Implemented Ng et al.'s dissimilarity measure

### PR DESCRIPTION
**Pull request by benandow**
Implemented Ng et al.'s dissimilarity measure[1] including necessary modifications to the base libraries. Feel free to pull into the main branch.

[1] Michael K. Ng, Mark Junjie Li, Joshua Zhexue Huang, and Zengyou He, "On the Impact of Dissimilarity Measure in k-Modes Clustering Algorithm", IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. 29, No. 3, January, 2007

### CODE REVIEW

**Comment by nicodv** [comment]
This is awesome, thank you!

I have 2 remarks, though:
- Do we need separate dissimilarity measures between the initialization and the main loop of the algorithm? I think it makes the API less clear, and I don't see the added value.
- The new dissimilarity measure needs to be properly unit tested before we can merge this.

**Comment by benandow** [comment]
Not a problem! It's a very useful library. Thanks for releasing it!

To address your remarks:
+ That's a fair point. I'll revert it to use the same dissimilarity measure for the initialization and main loop. The main reason why I implemented it with different parameters for dissimilarity measures is that it is more efficient to just invoke the simple matching_diss measure for initialization, as Ng. et al.'s measure inherently produces the same result due to the clusters being empty at that point. 
    + <i>Explanation</i>: The benefit of Ng et al.'s dissimilarity measure is that it essentially accounts for cluster impurity by penalizing attribute mismatches within a cluster. When the attribute value of the current record x<sub>i,j</sub> matches the corresponding attribute value of the cluster centroid z<sub>l,j</sub>, you look at the current members of the cluster and calculate the phi(z<sub>l,j</sub>, x<sub>i,j</sub>) as 1 - |C<sub>l,j,r</sub>| / |C<sub>l</sub>}|, where |C<sub>l</sub>| is the number of objects in the l<sup>th</sup> cluster, and  |C<sub>l,j,r</sub>| is the number of objects in the l<sup>th</sup> cluster whose j<sup>th</sup> attribute has the same attribute value of the current record x<sub>i,j</sub>. If the cluster is empty and the attribute value of the centroid matches the attribute value of the record (i.e., x<sub>i,j</sub> == z<sub>l,j</sub>), then there should be no penalty (i.e., 1 - |C<sub>l,j,r</sub>| / |C<sub>l</sub>}| = 0, as both |C<sub>l,j,r</sub>| and |C<sub>l</sub>}| equals 0). If x<sub>i,j</sub> != z<sub>l,j</sub>), then the penalty is simply 1.0. Therefore, when the cluster are empty, such as during centroid initialization, Ng. et al.'s dissimilarity measure produces the same result as the simple matching_diss measure.
+ I'll try to find some free time to include unit tests.

Best,
-Ben

**Comment by benandow** [comment]
As discussed above, I'll reverted k-modes and k-prototypes to use the same dissimilarity measure for initialization and the main loop. I added unit tests for the Ng dissimilarity function, and also error checking in the Ng dissimilarity measure to ensure that the 'membship' array has the correct shape.

File: kmodes/util/dissim.py [hunk 1]
```diff
@@ -7,14 +7,38 @@
...
+        raise ValueError("'membship' must be a rectangular array where the number of rows in 'membship' equals the number of rows in 'a' and the number of columns in 'membship' equals the number of rows in 'X'.")
+
+    return np.array([ np.array([calc_dissim(b, X, membship[idj], idr, idj) if b[idr] == t else 1.0 for idr,t in enumerate(val_a) ]).sum(0) for idj,val_a in enumerate(a) ])
+
+# Num objects w/ category value x_{i,r} for rth attr in jth cluster
+def calcCJR(b, X, memj, idr):
+    xcids = np.where(np.in1d(memj.ravel(), [1]).reshape(memj.shape))
+    return float((np.take(X, xcids, axis=0)[0][:, idr] == b[idr]).sum(0))
+
+def calc_dissim(b, X, memj, idr, idj):
```
Comment (nicodv) [inline kmodes/util/dissim.py]: `idj` is not used in the function at all?

Comment (benandow) [inline kmodes/util/dissim.py]: Good catch! I removed that parameter. I had implemented support for a weighting function, which was removed for the PR and that parameter was a leftover remnant.

File: kmodes/kmodes.py [hunk 2]
```diff
@@ -373,6 +376,9 @@ def fit_predict(self, X, y=None, **kwargs):
         """
         return self.fit(X, **kwargs).labels_
 
+    def genMembshipArray(self):
```
Comment (nicodv) [inline kmodes/kmodes.py]: Let's make this a function in utils/__init__.py

File: kmodes/kprototypes.py [hunk 3]
```diff
@@ -415,6 +416,9 @@ def fit(self, X, y=None, categorical=None):
                                                     self.verbose)
         return self
 
+    def genMembshipArray(self):
```
Comment (nicodv) [inline kmodes/kprototypes.py]: Let's make this a function in utils/__init__.py

**Comment by nicodv** [comment]
@benandow , in test_kmodes and test_kprototypes, we need to add tests that use this dissimilarity measure to prove that the end-to-end flow works, and that the results are as expected.

File: kmodes/kprototypes.py (hunk 3 above)
Comment (benandow) [inline kmodes/kprototypes.py]: Moved it and updated the references.

File: kmodes/kmodes.py (hunk 2 above)
Comment (benandow) [inline kmodes/kmodes.py]: Moved it and updated the references.

**Comment by benandow** [comment]
@nicodv I added the test cases in test_kmodes and test_kprototypes to show that the end-to-end flow works and the results are expected when using the Ng dissimilarity measure. 

After testing, I also submitted a bug fix that occurred when invoking fit and then trying to predict new data. I overlooked that we need to keep track of initial encoded feature vector for the attribute frequency calculation in Ng's dissimilarity measure. I implemented the patch. However, this means that the encoded FV will not be freed by the GC. An alternative solution is to build a map of attribute frequencies at the beginning, so that we can allow the encoded FV to be freed, but this adds unnecessary memory overhead if not predicting new data (e.g., if you're just invoking fit or fit_predict). Therefore, I believe that maintaining a reference to the encoded FV is the best choice.

**Comment by nicodv** [comment]
@benandow , I'm revisiting this presently.

I don't think we should save the (potentially very large) training matrix as `self.X`. Instead, I'd prefer to use the prediction matrix `X` in the predict method for calculating cost. At predict time, it's just for calculating cost anyway, so not really crucial.

**Comment by nicodv** [comment]
I went with using simple matching dissimilarity when predicting. A warning is printed for the user so that he/she is aware that Ng's dissimilarity is not used when predicting.

@benandow , thanks again for your contribution!

**Pull request merged**